design/baremetal-operator/configmap_for_host_inspection_data.md

```diff
@@ -0,0 +1,193 @@
+
+## Summary
+
+Introduce HostData Custom Resource (CR) to store host inspection data.
+
+## Motivation
+
+Currently inspection data is written to baremetalhost status subresource. When
+pivoting to a target cluster, with *`clusterctl move`*, status subresource is
+not copied with the host, because from Cluster-API perspective, it's controllers'
+job to rebuild the status in the target cluster. Also, according to the [Kubernetes API convention](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status), status is expected to be reconstructed by controllers if
+it's lost for any reason. Baremetal Operator (BMO) is capable of reconstructing
+host status by re-inspecting the host, but it will force the already provisioned
+host to go through another round of provisioning, which eventually reboots the
+host after IPA steps are completed. To have the inspection data available in the
+target cluster and to avoid unnecessary provisioning after pivoting, we currently
+copy inspection data to baremetalhost.metal3.io/status annotation so that BMO in
+the target cluster can continue reconciliation from a point it was before pivoting,
+avoiding redundant provisioning.
+
+However, there is a limitation with the current approach, because when inspection
+data is very large (e.g. when the host has 1000 multipath disks), it is easy to hit
+the global limit of `metadata.annotations` [262144](https://github.com/kubernetes/kubernetes/blob/master/test/integration/controlplane/synthetic_controlplane_test.go#L313) bytes and fail to pivot successfully.
+
+### Goals
+
+- Store host inspection data into Spec HostData CR
+- Adjust pivoting workflow to restore inspection data from HostData
+- Make the HostData CR immutable
+
+### Non-Goals
+
+Remove `status.hardware` from BareMetalHost CR
+
+## Proposal
+
+### User Stories
+
+As an operator, from an ephemeral cluster I would like to pivot/move to a target
+cluster successfully.
+
+## Design Details
+
+Apart from writing hardware details into the status of BareMetalHost, write it to a
+HostData CR spec. BareMetalHost and HostData will be parent and child objects
+respectively, and this relationship is controlled via ownerReferences chain. As such,
+the same name & namespace will be used for HostData as its owning object BareMetalHost.
+The spec of HostData will contain the same fields & hardware information as currently
+under `status.hardware` of BareMetalHost.
+
+### Pivoting
+
+There are three options one can choose to pivot, objects excluded from Cluster
+API chain, to a target cluster:
+
+1. [`clusterctl.cluster.x-k8s.io/move-hierarchy`](https://github.com/kubernetes-sigs/cluster-api/blob/0e6428151f2a388b98d1e220a698b85940a5f290/cmd/clusterctl/api/v1alpha3/labels.go#L38) label should be set
+    on BareMetalHost CRD before issuing *`clusterctl move`* so that the entire hierarchy
+    (i.e. including HostData, Secrets and ConfigMaps) is pivoted.
```

Comment:
I am a bit unsure still if thats the right label although it seems from the link you shared it is. This would need some testing

Reply:
You are right, and I removed other alternatives not to complicate it here.